subsys/net/lib/shell/dhcpv4.c

```diff
@@ -0,0 +1,208 @@
+	} else {
+		PR("DHCPv4 server started on interface %d\n", idx);
+	}
+#else /* CONFIG_NET_DHCPV4_SERVER */
+	PR_INFO("Set %s to enable %s support.\n",
+		"CONFIG_NET_DHCPV4_SERVER", "DHCPv4 server");
+#endif /* CONFIG_NET_DHCPV4_SERVER */
+	return 0;
+}
+
+static int cmd_net_dhcpv4_server_stop(const struct shell *sh, size_t argc, char *argv[])
+{
+#if defined(CONFIG_NET_DHCPV4_SERVER)
+	struct net_if *iface = NULL;
+	int idx, ret;
+
+	idx = get_iface_idx(sh, argv[1]);
+	if (idx < 0) {
+		return -ENOEXEC;
+	}
+
+	iface = net_if_get_by_index(idx);
+	if (!iface) {
+		PR_WARNING("No such interface in index %d\n", idx);
+		return -ENOEXEC;
+	}
+
+	ret = net_dhcpv4_server_stop(iface);
+	if (ret == -ENOENT) {
+		PR_WARNING("DHCPv4 server is not running on interface %d\n", idx);
+	} else if (ret < 0) {
+		PR_ERROR("DHCPv4 server failed to stop on interface %d, error %d\n",
+			 idx, -ret);
+	} else {
+		PR("DHCPv4 server stopped on interface %d\n", idx);
+	}
+#else /* CONFIG_NET_DHCPV4_SERVER */
+	PR_INFO("Set %s to enable %s support.\n",
+		"CONFIG_NET_DHCPV4_SERVER", "DHCPv4 server");
+#endif /* CONFIG_NET_DHCPV4_SERVER */
+	return 0;
+}
+
+#if defined(CONFIG_NET_DHCPV4_SERVER)
+static const char *dhcpv4_addr_state_to_str(enum dhcpv4_server_addr_state state)
+{
+	switch (state) {
+	case DHCPV4_SERVER_ADDR_FREE:
+		return "FREE";
+	case DHCPV4_SERVER_ADDR_RESERVED:
+		return "RESERVED";
+	case DHCPV4_SERVER_ADDR_ALLOCATED:
+		return "ALLOCATED";
+	case DHCPV4_SERVER_ADDR_DECLINED:
+		return "DECLINED";
+	}
+
+	return "<UNKNOWN>";
+}
+
+static uint32_t timepoint_to_s(k_timepoint_t timepoint)
+{
+	k_timeout_t timeout = sys_timepoint_timeout(timepoint);
+
+	if (K_TIMEOUT_EQ(timeout, K_NO_WAIT)) {
+		return 0;
+	}
+
+	if (K_TIMEOUT_EQ(timeout, K_FOREVER)) {
+		return UINT32_MAX;
+	}
+
+	return k_ticks_to_ms_floor64(timeout.ticks) / 1000;
+}
+
+static void dhcpv4_lease_cb(struct net_if *iface,
+			    struct dhcpv4_addr_slot *lease,
+			    void *user_data)
+{
+	struct net_shell_user_data *data = user_data;
+	const struct shell *sh = data->sh;
+	int *count = data->user_data;
+	char expiry_str[] = "4294967295"; /* Lease time is uint32_t, so take
+					   * theoretical max.
+					   */
+
+	if (*count == 0) {
+		PR("            Address\t    State\tExpiry (sec)\n");
+	}
+
+	(*count)++;
+
+	if (lease->state == DHCPV4_SERVER_ADDR_DECLINED) {
+		snprintk(expiry_str, sizeof(expiry_str) - 1, "infinite");
+	} else {
+		snprintk(expiry_str, sizeof(expiry_str) - 1, "%u",
+			 timepoint_to_s(lease->expiry));
+	}
+
+	PR("%2d. %15s\t%9s\t%12s\n", *count, net_sprint_ipv4_addr(&lease->addr),
+	   dhcpv4_addr_state_to_str(lease->state), expiry_str);
+}
+#endif /* CONFIG_NET_DHCPV4_SERVER */
+
+static int cmd_net_dhcpv4_server_status(const struct shell *sh, size_t argc, char *argv[])
+{
+#if defined(CONFIG_NET_DHCPV4_SERVER)
+	struct net_shell_user_data user_data;
+	struct net_if *iface = NULL;
+	int idx, ret;
+	int count = 0;
+
+	idx = get_iface_idx(sh, argv[1]);
```

Comment:
One improvement here could be that if user does not supply interface, the code could call the foreach function with iface NULL and get information about all the interfaces. Could be done in future PR of course.

Reply:
That could be a good improvement, I often found myself forgetting the iface index when checking the status. Will do.

Reply:
Done, please re-check.